python/Homework.py

```diff
@@ -0,0 +1,210 @@
+# Scatter plot is easy to tell the difference if the difference is obvious among the kinds.
+
+# 3-2 Violin Plot
+
+df = px.data.tips()
+fig = px.violin(
+    Iris_df,
+    y="petal width in cm",
+    color="class",
+    violinmode="overlay",  # draw violins on top of each other
+    # default violinmode is 'group' as in example above
+    hover_data=Iris_df.columns,
+)
+fig.show()
+
+# Virginica has longest pedal width, but Senoto is lowest
+# Violin plot is very straightforward but it only describe one predictor in one plot.
+
+# 3-3 3D Scatter Plot
+
+fig = px.scatter_3d(
+    Iris_df,
+    x="sepal length in cm",
+    y="sepal width in cm",
+    z="petal length in cm",
+    color="petal width in cm",
+    symbol="class",
+)
+
+fig.show()
+
+# Virginica has higher sepeal width/lenghth and petal width/length than other two kinds.
+# 3D scatter Plot can discribe more characters than my first scatter plot
+
+# 3-4 3D bubble
+
+fig = px.scatter_3d(
+    Iris_df,
+    x="sepal length in cm",
+    y="sepal width in cm",
+    z="petal length in cm",
+    size="petal width in cm",
+    color="class",
+    hover_data=["class"],
+)
+
+fig.show()
+# It is similar with 3D Scatter plot.
+
+# 3-5 Polar Charts
+
+fig = px.scatter_polar(
+    Iris_df, r="petal length in cm", theta="class", size="petal width in cm"
+)
+fig.show()
+
+# Viginica's petal length and width are longer than other two.
+# These three kinds are distributed in different radius size round.
+# so it is easy to see how much difference about petal among three different kinds
+
+# 3-6 PCA Visualization in Python
+
+fig = px.scatter_matrix(Iris_df, dimensions=heads, color="class")
+fig.update_traces(diagonal_visible=False)
+fig.show()
+
+# PCA Visulization can help us to check the distribution of three kinds under different characters.
+# So based on the plot, we can know Virginica always has the longest length and widest width for Sepal and petal
+
+# 4 Use the StandardScaler transformer
+# 4-1 DataFrame to numpy values
+
+X_orig = Iris_df[
+    [
+        "sepal length in cm",
+        "sepal width in cm",
+        "petal length in cm",
+        "petal width in cm",
+    ]
+].values
+y = Iris_df["class"].values
+
+# 4-2 Generate a feature from the where they started
+
+scaler = StandardScaler()
+scaler.fit(X_orig)
+X = scaler.transform(X_orig)
+
+# 4-3 Fit the features to a random forest
+
+random_forest = RandomForestClassifier(random_state=100)
+random_forest.fit(X, y)
+
+test_df = Iris_df.iloc[:, 0:4]
+
+print(test_df)
+
+X_test_orig = test_df.values
+X_test = scaler.transform(X_test_orig)
+prediction = random_forest.predict(X_test)
+probability = random_forest.predict_proba(X_test)
+
+print(f"Classes: {random_forest.classes_}")
+print(f"Probability: {probability}")
+print(f"Predictions: {prediction}")
+
+# 4-4 Fit the features to a Decision Tree
+
+Decision_tree = tree.DecisionTreeClassifier(random_state=12)
+Decision_tree.fit(X, y)
+
+test_df = Iris_df.iloc[:, 0:4]
```

Comment:
Line 169-174: You already defined these variables in section 4-3 above.

Reply:
Yes, I delete them! Thank you!